from os.path import join
import pandas as pd
from snakemake.utils import min_version
##### set minimum snakemake version #####
min_version("5.1.2")

##### load config and sample sheets #####
configfile: "config/config.yaml"
#validate(config, schema="schemas/config.schema.yaml")

patients = pd.read_csv(config["patients"], sep="\t").set_index("patient", drop=False)
# validate(samples, schema="schemas/samples.schema.yaml")

samples = pd.read_csv(config["samples"], dtype=str, sep="\t").set_index(["patient", "sample"], drop=False)
# print(samples)
samples.index = samples.index.set_levels([i.astype(str) for i in samples.index.levels])  # enforce str in index
# validate(units, schema="schemas/units.schema.yaml")
samples = samples.loc[samples.infection == "Salmonella enterica serovar Typhimurium"]

# urls
BASE_SL1344_URL = "https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/210/855/GCF_000210855.2_ASM21085v2"

# Directories
FASTQ_DIR = "FASTQ"
ENV_DIR = join("..", "envs")

# Files
STAR_ENV_FILE = join(ENV_DIR, "star.yml")

# Input Files
SRA_FASTQ1_FILE = join(FASTQ_DIR, "{sample}_1.fastq")
SRA_FASTQ2_FILE = join(FASTQ_DIR, "{sample}_2.fastq")
HUMAN_FA = "/fdb/GATK_resource_bundle/hg38bundle/Homo_sapiens_assembly38.fasta"
HUMAN_GTF = "raw/genome/gencode.v22.annotation.gtf"

PATIENT_FASTQ_DIR = join("FASTQ", "raw", "{patient}")
CR_SAMPLE_ODIR = join("output", "CellRanger", "{patient}-{sample}")

CB_FASTQ_FILE = join(PATIENT_FASTQ_DIR, "{sample}_S1_{lane}_R1_001.fastq.gz")
cDNA_FASTQ_FILE = join(PATIENT_FASTQ_DIR, "{sample}_S1_{lane}_R2_001.fastq.gz")
IDX_FASTQ_FILE = join(PATIENT_FASTQ_DIR, "{sample}_S1_{lane}_I1_001.fastq.gz")
CR_BAM_FILE = join(CR_SAMPLE_ODIR, "outs", "possorted_genome_bam.bam")
# Intermediate Files
# cDNA_FASTQ_FILE = join(FASTQ_DIR, "{patient}-{sample}",
#                        "cDNAfragmentSequence.fastq.gz")
# CellBarcode_FASTQ_FILE = join(FASTQ_DIR, "{patient}-{sample}",
#                               "CellBarcodeUMIsequence.fastq.gz")
#BAM_FILE = join("output", "HG38-SL1344", "STARsolo", "{patient}-{sample}", "Aligned.sortedByCoord.out.bam")
#SL1344_FA = join("raw", "genome", "SL1344_genome.fa")
#SL1344_GTF = join("raw", "genome", "SL1344.gtf")

# Output File
# FILTERED_PATHSEQ_READS_FILE = join(SC_PATHSEQ_OUTPUT, "Filtered_PathSeq_reads.tsv")

# Snakemake includes
include: "../../RNA-snakemake-rules/rules/genome.smk"
# include: "../../RNA-snakemake-rules/rules/STARsolo.smk"
# include: "../../pathogen-discovery-rules/rules/PathSeq.smk"

rule all:
    input:
        expand(CB_FASTQ_FILE, zip, patient=samples["patient"], sample=samples["sample"], lane=samples["lane"])
        # join(config["STARsolo"]["output_dir"], "Aligned.sortedByCoord.out.bam").format(patient="Pt0", sample="infected"),



# # special rules for STARsolo since we want to combine sequencing runs
# rule create_star_index:
#     conda:
#         STAR_ENV_FILE
#     input:
#         config["ref"]["genome"]
#     output:
#         directory(config["STARsolo"]["genome_index"])
#     threads:
#         16
#     shell:
#         "mkdir '{output}' && STAR "
#         "--runThreadN {threads} "
#         "--runMode genomeGenerate "
#         "--genomeDir '{output}' "
#         "--genomeFastaFiles '{input}'"

# # STARsolo
# rule run_STARsolo_10x:
#     conda:
#         STAR_ENV_FILE
#     input:
#         barcode_whitelist = config["ref"]["barcode_whitelist"],
#         cDNA_fq = expand(cDNA_FASTQ_FILE, patient=samples["patient"], sample=samples["sample"]),
#         CellBarcode_fq = expand(CellBarcode_FASTQ_FILE, patient=samples["patient"], sample=samples["sample"]),
#         index = config["STARsolo"]["genome_index"],
#         gtf = config["ref"]["annotation"]
#     output:
#         join(config["STARsolo"]["output_dir"], "Aligned.sortedByCoord.out.bam").format(patient="Pt0", sample="infected"),
#         join(config["STARsolo"]["output_dir"], "Solo.out", "Gene", "filtered", "matrix.mtx").format(patient="Pt0", sample="infected")
#     params:
#         odir = join(config["STARsolo"]["output_dir"], "").format(patient="Pt0", sample="infected")
#     threads:
#         48
#     shell:
#         "STAR "
#         "--runThreadN {threads} "
#         "--genomeDir '{input.index}' "
#         "--genomeLoad NoSharedMemory "
#         "--outFileNamePrefix '{params.odir}' "
#         "--readFilesIn '{input.cDNA_fq[0]},{input.cDNA_fq[1]},{input.cDNA_fq[2]},{input.cDNA_fq[3]}' "
#         "'{input.CellBarcode_fq[0]},{input.CellBarcode_fq[1]},{input.CellBarcode_fq[2]},{input.CellBarcode_fq[3]}' "
#         "--soloType CB_UMI_Simple "
#         "--soloCBwhitelist {input.barcode_whitelist} "
#         "--soloFeatures Gene "
#         "--outSAMattrRGline ID:{wildcards.patient}.{wildcards.sample} "
#         "PL:illumina SM:{wildcards.patient}.{wildcards.sample} LB:RNA "
#         # CR/UR is raw (uncorrected) CellBarcode/UMI while CB/UB is corrected
#         "--outSAMattributes NH HI AS nM NM ch CR UR CB UB "
#         "--outSAMtype BAM SortedByCoordinate "  # required for CB/UB
#         "--outSAMunmapped Within "  # include the unmapped reads in the output file
#         "--readFilesCommand zcat "  # used for reading gz files
#         "--sjdbGTFfile '{input.gtf}' "
#         + config["params"]["STARsolo"]


# # create combined genome
# rule combine_genome:
#     group:
#         "reference"
#     input:
#         HUMAN_FA,
#         SL1344_FA
#     output:
#         config["ref"]["genome"]
#     shell:
#         "cat {input[0]} {input[1]} > {output}"
#
# rule combine_annotation:
#     group:
#         "reference"
#     input:
#         HUMAN_GTF,
#         SL1344_GTF
#     output:
#         config["ref"]["annotation"]
#     shell:
#         "cat {input[0]} {input[1]} > {output}"
#
# # download Salmonella SL1344 genome and gtf file
# rule download_SL1344_files:
#     group:
#         "reference"
#     params:
#         GTF_URL=join(BASE_SL1344_URL, "GCF_000210855.2_ASM21085v2_genomic.gtf.gz"),
#         FA_URL=join(BASE_SL1344_URL, "GCF_000210855.2_ASM21085v2_genomic.fna.gz")
#     output:
#         FA=SL1344_FA,
#         GTF=SL1344_GTF,
#     shell:
#         "wget -O - {params.FA_URL} | gunzip -c > {output.FA} && "
#         "wget -O - {params.GTF_URL} | gunzip -c > {output.GTF}"

# download files from SRA
# need to --include-technical to get the 10x UMI and CB reads
# this requires
rule download_FASTQ_from_SRA:
    group:
        "FASTQ"
    params:
        run = lambda wildcards: samples.at[(wildcards.patient, wildcards.sample, wildcards.lane), "Run"],
        dir = FASTQ_DIR
    output:
        temp(SRA_FASTQ1_FILE),
        temp(SRA_FASTQ2_FILE),
        temp(SRA_FASTQ3_FILE),
    shell:
        "module load sratoolkit && "
        "fasterq-dump -O {params.dir} -t /lscratch/$SLURM_JOBID "
        "--include-technical --split-files {params.run}"

# compress and rename
rule compress_rename_SRA_FASTQ_files:
    group:
        "FASTQ"
    input:
        CellBarcode_fq = SRA_FASTQ1_FILE,
        cDNA_fq = SRA_FASTQ2_FILE,
        idx_fq = SRA_FASTQ3_FILE
    output:
        CellBarcode_fq = CB_FASTQ_FILE,
        cDNA_fq = cDNA_FASTQ_FILE,
        idx_fq = IDX_FASTQ_FILE
    shell:
        "gzip -vc {input.CellBarcode_fq} > {output.CellBarcode_fq} && "
        "gzip -vc {input.cDNA_fq} > {output.cDNA_fq} && "
        "gzip -vc {input.idx_fq} > {output.idx_fq}"
