from os.path import join

SRPRISM_DB_DIR = join("output", "genomes_db")
GENOME_DB_PREFIX = join(SRPRISM_DB_DIR, "RNA-spikein-sequences")
GENOME_DB_FILE = join(SRPRISM_DB_DIR, "RNA-spikein-sequences.idx")

# file containing sequences for RNA Spike In sequences
RNA_SPIKE_SEQUENCES = join("data", "E-MTAB-5466_spike_in_sequences.fa")

#
rule filter_spike_reads_from_PathSeq_bam:
    input:
        join("..", "output", "PathSeq", "BC06-BC06-BC06", "pathseq.bam"),
        join("output", "BC06-BC06-BC06", "excluded_reads.txt")
    output:
        join("..", "output", "PathSeq", "BC06-BC06-BC06", "pathseq.filtered.bam"),
    shell:
        "module load picard && "
        "java -jar $PICARDJARPATH/picard.jar FilterSamReads "
        "I={input[0]} O={output[0]} "
        "FILTER=excludeReadList READ_LIST_FILE={input[1]}"


rule extract_spike_querynames:
    input:
        join("output", "SRPRISM", "BC06-BC06-BC06", "paired.sam"),
        join("output", "SRPRISM", "BC06-BC06-BC06", "unpaired.sam")
    output:
        join("output", "BC06-BC06-BC06", "excluded_reads.txt")
    shell:
        "module load samtools && "
        "samtools view {input[0]} | cut -f 1 > {output} && "
        "samtools view {input[1]} | cut -f 1 >> {output}"


rule map_paired_reads_against_RNA_SPIKE_SEQUENCES:
    input:
        join("output", "PathSeq", "BC06-BC06-BC06", "paired_1.fastq"),
        join("output", "PathSeq", "BC06-BC06-BC06", "paired_2.fastq"),
        GENOME_DB_FILE
    output:
        join("output", "SRPRISM", "BC06-BC06-BC06", "paired.sam")
    shell:
        "srprism search -I {params} -i {input[0]},{input[1]} -F fastq -p true -o {output} --sam-header true"

rule map_SRPRISM_genome_unpaired:
    params:
        GENOME_DB_PREFIX
    input:
        join("output", "PathSeq", "BC06-BC06-BC06", "unpaired.fastq"),
        GENOME_DB_FILE
    output:
        join("output", "SRPRISM", "BC06-BC06-BC06", "unpaired.sam")
    shell:
        "srprism search -I {params} -i {input[0]} -F fastq -p false -o {output} --sam-header true"

rule convert_paired_bam_to_fastq:
    input:
        join("..", "output", "PathSeq", "BC06-BC06-BC06", "pathseq.bam")
    output:
        join("output", "PathSeq", "BC06-BC06-BC06", "paired_1.fastq"),
        join("output", "PathSeq", "BC06-BC06-BC06", "paired_2.fastq"),
        join("output", "PathSeq", "BC06-BC06-BC06", "unpaired.fastq"),
    shell:
        "module load picard && "
        "java -jar $PICARDJARPATH/picard.jar SamToFastq "
        "I={input} F={output[0]} F2={output[1]} FU={output[2]}"

# rule extract_paired_read_BAM:
#     input:
#         join("output", "PathSeq", "BC06-BC06-BC06", "pathseq.sorted.bam")
#     output:
#         join("output", "PathSeq", "BC06-BC06-BC06", "pathseq.paired.bam")
#     shell:
#         "module load samtools && "
#         "samtools view -h -b -f 0x1 {input} > {output}"
#
# rule sort_PathSeq_BAM_by_queryname:
#     input:
#         join("..", "output", "PathSeq", "BC06-BC06-BC06", "pathseq.bam")
#     output:
#         join("output", "PathSeq", "BC06-BC06-BC06", "pathseq.sorted.bam")
#     shell:
#         "module load picard && "
#         "java -jar $PICARDJARPATH/picard.jar SortSam "
#         "I={input} O={output} SORT_ORDER=queryname"

rule make_reference_DB:
    params:
        GENOME_DB_PREFIX
    input:
        RNA_SPIKE_SEQUENCES
    output:
        GENOME_DB_FILE
    shell:
        "srprism mkindex -i {input} -o {params}"


rule download_RNA_spikes:
    params:
        url="https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-5466/E-MTAB-5466.additional.1.zip",
        dir="data"
    output:
        RNA_SPIKE_SEQUENCES,
        temp("temp.zip"),
        temp(join("data", "E-MTAB-5466_spike_in_concentration.txt"))
    shell:
        "wget {params.url} -O temp.zip && unzip temp.zip -d {params.dir}"
