from os.path import join, basename
import pandas as pd
from snakemake.utils import min_version
##### set minimum snakemake version #####
min_version("5.1.2")
##### load config and sample sheets #####
configfile: join("config", "PathSeq-config.yaml")
#validate(config, schema="schemas/config.schema.yaml")

patients = pd.read_csv(config["patients"], sep="\t").set_index("patient", drop=False)
# validate(samples, schema="schemas/samples.schema.yaml")

samples = pd.read_csv(config["samples"], dtype=str, sep="\t").set_index(["patient", "sample"], drop=False)
# print(samples)
samples.index = samples.index.set_levels([i.astype(str) for i in samples.index.levels])  # enforce str in index

FASTQ1_FILE = join("FASTQ", "raw", "{patient}-{sample}_1.fastq.gz")
FASTQ2_FILE = join("FASTQ", "raw", "{patient}-{sample}_2.fastq.gz")

include: "../../RNA-snakemake-rules/rules/mouse-genome.smk"
include: "../../RNA-snakemake-rules/rules/STAR.smk"
include: "../../pathogen-discovery-rules/rules/build-PathSeq-host-files.smk"
include: "../../pathogen-discovery-rules/rules/PathSeq.smk"
# steps

rule all:
    input:
        expand(join("output", "PathSeq", "{patient}-{sample}", "pathseq.bam"), zip, sample=samples["sample"], patient=samples["patient"])

# move data so in FASTQ/raw/{wildcards.patient}-{wildcards.sample}_1.fastq.gz format for STAR
rule get_FASTQ_files:
    params:
        "/data/Robinson-SB/Seq1_HFJLTBGXC_SS2_NoLaneSplit/{sample}_S*_R1_001.fastq.gz",
        "/data/Robinson-SB/Seq1_HFJLTBGXC_SS2_NoLaneSplit/{sample}_S*_R2_001.fastq.gz"
    output:
        temp(FASTQ1_FILE),
        temp(FASTQ2_FILE)
    shell:
        "cp {params[0]} {output[0]} && "
        "cp {params[1]} {output[1]}"
