from os.path import join
import pandas as pd
from snakemake.utils import min_version
##### set minimum snakemake version #####
min_version("5.1.2")

#patient = config["patient"]
##### load config and sample sheets #####
configfile: join("config", "PathSeq-config.yaml")
patients = pd.read_csv(config["patients"], sep="\t").set_index("patient", drop=False)
samples = pd.read_csv(config["samples"], dtype=str, sep="\t").set_index(["patient", "sample"], drop=False)
samples.index = samples.index.set_levels([i.astype(str) for i in samples.index.levels])  # enforce str in index
samples = samples.loc[samples["patient"] == "P0104"]
# directories


pathseq_bam = join("output", "PathSeq", "{patient}-{sample}", "pathseq.bam")

# include rules

include: "../pathogen-discovery-rules/rules/PathSeq-single.smk"


rule all:
    input:
        expand(pathseq_bam, zip, sample=samples["sample"], patient=samples["patient"])


rule build_host_kmer_file:
    input:
        config["human_ref"]["genome"]
    output:
        config["PathSeq"]["host_bfi"]
    shell:
        "module load GATK/4.1.8.1 && "
        "gatk PathSeqBuildKmers "
        "--java-options '-Xmx80g' "
        "--reference '{input}' "
        "-O '{output}'"

  host_img: "output/PathSeq/host.fasta.img"
  host_bfi

rule build_host_BWA_image:
    input:
        config["human_ref"]["genome"]
    output:
        config["PathSeq"]["host_img"]
    shell:
        "module load GATK/4.1.8.1 && "
        "gatk BwaMemIndexImageCreator -I {input} -O {output}"
