from os.path import join, basename
import pandas as pd
from snakemake.utils import min_version
##### set minimum snakemake version #####
min_version("5.1.2")
# patient = config["patient"]
# malignant = config["celltype"]
# start = config["start"]
# stop = config["stop"]
##### load config and sample sheets #####
configfile: join("config", "PathSeq-config.yaml")
#validate(config, schema="schemas/config.schema.yaml")

patients = pd.read_csv(config["patients"], sep="\t").set_index("patient", drop=False)
# validate(samples, schema="schemas/samples.schema.yaml")

samples = pd.read_csv(config["samples"], dtype=str, sep="\t").set_index(["patient", "sample"], drop=False)
# print(samples)
samples.index = samples.index.set_levels([i.astype(str) for i in samples.index.levels])  # enforce str in index
plates_of_interest = [
    "cy53_1_CD45_neg", "Cy71_CD45", "CY80_neg", "Cy80_II_CD45", "cy80_CD45_neg",
    "Cy81_Bulk_CD45", "Cy81_FNA_CD45", "cy82_CD45_pos_1", "cy82_CD45_pos_2",
    "cy82_CD45_pos_3"
]
samples = samples.loc[samples.plate.isin(["cy53_1_CD45_neg"])]
# samples = samples.loc[samples["patient"] == "Melanoma_53"]
print(samples)
# directories
FASTQ_DIR = join("FASTQ", "raw")

# Intermediate Files
FASTQ1_FILE = join(FASTQ_DIR,  "{patient}-{sample}_1.fastq.gz")
FASTQ2_FILE = join(FASTQ_DIR, "{patient}-{sample}_2.fastq.gz")
TRIMMED_FASTQ1_FILE = join("FASTQ", "trimmed", "{patient}-{sample}_1.fastq.gz")
TRIMMED_FASTQ2_FILE = join("FASTQ", "trimmed", "{patient}-{sample}_2.fastq.gz")
TRIMMED_UNPAIRED_FILE = join("FASTQ", "trimmed", "{patient}-{sample}_3.fastq.gz")

# Output Files
PATHSEQ_BAM = join("output", "PathSeq", "{patient}-{sample}", "pathseq.bam")

include: "../RNA-snakemake-rules/rules/genome.smk"
include: "../RNA-snakemake-rules/rules/trim.smk"
include: "../pathogen-discovery-rules/rules/filter-host-reads.smk"
include: "../pathogen-discovery-rules/rules/PathSeq-host-filter-batch.smk"


rule all:
    input:
        expand(config["PathSeq"]["bam_file"], zip, patient=samples["patient"], sample=samples["sample"])


rule download_FASTQ_files:
    group:
        "FASTQ"
    params:
        fq1 = lambda wildcards: samples.at[(wildcards.patient, wildcards.sample), "fq1_url"],
        fq2 = lambda wildcards: samples.at[(wildcards.patient, wildcards.sample), "fq2_url"],
    output:
        fq1=temp(FASTQ1_FILE),
        fq2=temp(FASTQ2_FILE)
    shell:
        "module load google-cloud-sdk && "
        "gsutil cp {params.fq1} {output.fq1}  && "
        "gsutil cp {params.fq2} {output.fq2}"
