from os.path import join
import pandas as pd

# urls
LT2_genome_URL = "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/945/GCF_000006945.2_ASM694v2/GCF_000006945.2_ASM694v2_genomic.fna.gz"
D23580_genome_URL = "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/900/538/085/GCF_900538085.1_D23580_liv/GCF_900538085.1_D23580_liv_genomic.fna.gz"
# files
GENOME_FA = join("raw", "{genome}.fa")
SRPRISM_DB_DIR = join("output", "genome_db")
GENOME_DB_PREFIX = join(SRPRISM_DB_DIR, "{genome}")
GENOME_DB_FILE = join(SRPRISM_DB_DIR, "{genome}.idx")

Input_BAM = join("..", "output", "STAR", "{patient}-{sample}", "host.reads.filtered.bam")
Input_FQ1 = join("output", "{patient}-{sample}", "host.reads.filtered_1.fq")
Input_FQ2 = join("output", "{patient}-{sample}", "host.reads.filtered_2.fq")
Input_FQ3 = join("output", "{patient}-{sample}", "host.reads.filtered_3.fq")
SRPRISM_PAIRED_SAM = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-paired.sam")
SRPRISM_UNPAIRED_SAM = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-unpaired.sam")
SRPRISM_PAIRED_BAM = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-paired.bam")
SRPRISM_UNPAIRED_BAM = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-unpaired.bam")
SRPRISM_SORTED_PAIRED_BAM = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-sorted_paired.bam")
SRPRISM_SORTED_UNPAIRED_BAM = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-sorted_unpaired.bam")
SRPRISM_BAM = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-merged.bam")
SRPRISM_BAI = join("output", "SRPRISM", "{patient}-{sample}", "{genome}-merged.bai")


samples = pd.read_csv("../data/samples.tsv", dtype=str, sep="\t")
LT2_infected_samples = samples.loc[(samples.infection == "LT2") & (samples.status == "infected")]
D23580_infected_samples = samples.loc[(samples.infection == "D23580") & (samples.status == "infected")]
LT2_infected_samples = LT2_infected_samples.iloc[0:5]

rule all:
    input:
        expand(SRPRISM_BAI, zip, patient=LT2_infected_samples["patient"], sample=LT2_infected_samples["sample"], genome=LT2_infected_samples["infection"]),
        #expand(SRPRISM_BAI, zip, patient=D23580_infected_samples["patient"], sample=D23580_infected_samples["sample"], genome=D23580_infected_samples["infection"]),

rule create_BAM_index:
    input:
        SRPRISM_BAM
    output:
        SRPRISM_BAI
    shell:
        "module load samtools && "
        "samtools index {input} {output}"

rule merged_BAM_files:
    input:
        SRPRISM_SORTED_PAIRED_BAM,
        SRPRISM_SORTED_UNPAIRED_BAM
    output:
        SRPRISM_BAM
    shell:
        "module load samtools && "
        "samtools merge {output} {input[0]} {input[1]}"

rule convert_unpaired_SAM_to_sorted_BAM:
    input:
        SRPRISM_UNPAIRED_SAM
    output:
        SRPRISM_UNPAIRED_BAM,
        SRPRISM_SORTED_UNPAIRED_BAM
    shell:
        "module load samtools && "
        "samtools view -bS {input} > {output[0]} && "
        "samtools sort -o {output[1]} {output[0]}"

rule convert_paired_SAM_to_sorted_BAM:
    input:
        SRPRISM_PAIRED_SAM
    output:
        SRPRISM_PAIRED_BAM,
        SRPRISM_SORTED_PAIRED_BAM
    shell:
        "module load samtools && "
        "samtools view -bS {input} > {output[0]} && "
        "samtools sort -o {output[1]} {output[0]}"

rule map_SRPRISM_genome_unpaired:
    params:
        GENOME_DB_PREFIX
    input:
        Input_FQ3,
        GENOME_DB_FILE
    output:
        SRPRISM_UNPAIRED_SAM
    shell:
        "srprism search -I {params} -i {input[0]} -F fastq -p false -o {output} --sam-header true"

rule map_SRPRISM_genome_paired:
    params:
        GENOME_DB_PREFIX
    input:
        Input_FQ1,
        Input_FQ2,
        GENOME_DB_FILE
    output:
        SRPRISM_PAIRED_SAM
    shell:
        "srprism search -I {params} -i {input[0]},{input[1]} -F fastq -p true -o {output} --sam-header true"

rule convert_BAM_to_paired_FASTQ:
    input:
        Input_BAM
    output:
        Input_FQ1,
        Input_FQ2,
        Input_FQ3
    shell:
        "module load picard && "
        "java -jar $PICARDJARPATH/picard.jar SamToFastq I={input} "
        "F={output[0]} F2={output[1]} FU={output[2]}"

rule make_SRPRISM_DB:
    params:
        GENOME_DB_PREFIX,
        GENOME_DB_DIR
    input:
        GENOME_FA
    output:
        GENOME_DB_FILE
    shell:
        #"mkdir {params[1]} && "
        "srprism mkindex -i {input} -o {params[0]}"

rule download_D23580_genome:
    wildcard_constraints:
        genome="D23580"
    params:
        D23580_genome_URL
    output:
        GENOME_FA
    shell:
        "wget -O - {params} | gunzip -c > {output} "

rule download_LT2_genome:
    wildcard_constraints:
        genome="LT2"
    params:
        LT2_genome_URL
    output:
        GENOME_FA
    shell:
        "wget -O - {params} | gunzip -c > {output} "

# wget -O - {params.FA_URL} | gunzip -c > {output.FA}
